src/main/java/org/jetbrains/bsp/bazel/server/bsp/services/ScalaBuildServerService.java

```diff
@@ -54,10 +63,55 @@ public ScalaBuildServerService(BazelData bazelData, BazelRunner bazelRunner) {
     return CompletableFuture.completedFuture(new ScalaTestClassesResult(new ArrayList<>()));
   }
 
-  public CompletableFuture<ScalaMainClassesResult> buildTargetScalaMainClasses(
+  public Either<ResponseError, ScalaMainClassesResult> buildTargetScalaMainClasses(
       ScalaMainClassesParams scalaMainClassesParams) {
-    System.out.printf("DWH: Got buildTargetScalaMainClasses: %s%n", scalaMainClassesParams);
-    // TODO(illicitonion): Populate
-    return CompletableFuture.completedFuture(new ScalaMainClassesResult(new ArrayList<>()));
+    TargetRulesResolver<ScalaMainClassesItem> targetRulesResolver =
+        TargetRulesResolver.withBazelRunnerAndMapper(bazelRunner, this::mapRuleToMainClassesItem);
+
+    List<ScalaMainClassesItem> resultItems =
+        targetRulesResolver.getItemsForTargets(scalaMainClassesParams.getTargets());
+
+    ScalaMainClassesResult result = new ScalaMainClassesResult(resultItems);
+
+    return Either.forRight(result);
+  }
+
+  private ScalaMainClassesItem mapRuleToMainClassesItem(Build.Rule rule) {
+    BuildTargetIdentifier targetId = new BuildTargetIdentifier(rule.getName());
+    List<ScalaMainClass> mainClasses = collectMainClasses(rule);
+    return new ScalaMainClassesItem(targetId, mainClasses);
+  }
+
+  private List<ScalaMainClass> collectMainClasses(Build.Rule rule) {
+    List<String> targetOptions =
+        getAttribute(rule, SCALA_COMPILER_OPTIONS_NAME)
+            .flatMap(attr -> attr.getStringListValueList().stream())
+            .collect(Collectors.toList());
```

Comment:
Isn't it the same case as the two methods extracted below, that is, could it be extracted to something like _getTargetOptions_?

Reply:
It's not exactly the same, two methods use getStringListValueList (because the attributes are of type string list) and one uses getStringValue. 
Anyway, I extracted the methods with the same logic